File: android/app/src/main/kotlin/com/namsoon/footballnote/MediaPipePoseLandmarkerChannel.kt
package com.namsoon.footballnote

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.ImageFormat
import android.graphics.Matrix
import android.graphics.Rect
import android.graphics.YuvImage
import android.os.Handler
import android.os.Looper
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarker
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.ByteArrayOutputStream
import java.util.Optional
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import kotlin.math.max
import kotlin.math.roundToInt

class MediaPipePoseLandmarkerChannel(
    private val context: Context,
    messenger: BinaryMessenger,
) : MethodChannel.MethodCallHandler {
    private val channel = MethodChannel(messenger, channelName)
    private val executor: ExecutorService = Executors.newSingleThreadExecutor()
    private val mainHandler = Handler(Looper.getMainLooper())
    private var poseLandmarker: PoseLandmarker? = null
    private var lastTimestampMs = 0L

    init {
        channel.setMethodCallHandler(this)
    }

    fun dispose() {
        channel.setMethodCallHandler(null)
        executor.execute {
            poseLandmarker?.close()
            poseLandmarker = null
        }
        executor.shutdown()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            detectMethodName -> detectPoseFromNv21(call, result)
            closeMethodName -> closeLandmarker(result)
            else -> result.notImplemented()
        }
    }

    private fun detectPoseFromNv21(call: MethodCall, result: MethodChannel.Result) {
        val bytes = call.argument<ByteArray>("bytes")
        val width = call.argument<Int>("width") ?: 0
        val height = call.argument<Int>("height") ?: 0
        val rotationDegrees = call.argument<Int>("rotationDegrees") ?: 0
        val timestampMs = call.argument<Number>("timestampMs")?.toLong() ?: 0L

        if (bytes == null || width <= 0 || height <= 0) {
            result.error(
                "invalid_frame",
                "MediaPipe pose detection requires NV21 bytes with width and height.",
                null,
            )
            return
        }

        executor.execute {
            try {
                val detection = detectPose(
                    bytes = bytes,
                    width = width,
                    height = height,
                    rotationDegrees = rotationDegrees,
                    timestampMs = timestampMs,
                )
                mainHandler.post { result.success(detection) }
            } catch (error: Exception) {
                mainHandler.post {
                    result.error(
                        "mediapipe_pose_failed",
                        error.message ?: "MediaPipe pose detection failed.",
                        null,
                    )
                }
            }
        }
    }

    private fun closeLandmarker(result: MethodChannel.Result) {
        executor.execute {
            poseLandmarker?.close()
            poseLandmarker = null
            mainHandler.post { result.success(null) }
        }
    }

    private fun detectPose(
        bytes: ByteArray,
        width: Int,
        height: Int,
        rotationDegrees: Int,
        timestampMs: Long,
    ): Map<String, Any?> {
        val sourceBitmap = nv21ToBitmap(bytes, width, height)
        val rotatedBitmap = rotateBitmap(sourceBitmap, rotationDegrees)
        if (rotatedBitmap !== sourceBitmap) {
            sourceBitmap.recycle()
        }
        val bitmap = resizeBitmapIfNeeded(rotatedBitmap)
        if (bitmap !== rotatedBitmap) {
            rotatedBitmap.recycle()
        }

        try {
            val mpImage = BitmapImageBuilder(bitmap).build()
            val safeTimestampMs = nextTimestamp(timestampMs)
            val result = landmarker().detectForVideo(mpImage, safeTimestampMs)
            val landmarks = result.landmarks().firstOrNull().orEmpty()
            val worldLandmarks = result.worldLandmarks().firstOrNull().orEmpty()
            val encodedLandmarks = landmarks.mapIndexed { index, landmark ->
                val world = worldLandmarks.getOrNull(index)
                mapOf(
                    "index" to index,
                    "x" to (landmark.x() * bitmap.width),
                    "y" to (landmark.y() * bitmap.height),
                    "z" to landmark.z(),
                    "visibility" to optionalFloat(landmark.visibility()),
                    "presence" to optionalFloat(landmark.presence()),
                    "worldX" to world?.x(),
                    "worldY" to world?.y(),
                    "worldZ" to world?.z(),
                    "worldVisibility" to world?.visibility()?.let(::optionalFloat),
                )
            }
            return mapOf(
                "imageWidth" to bitmap.width,
                "imageHeight" to bitmap.height,
                "landmarks" to encodedLandmarks,
            )
        } finally {
            bitmap.recycle()
        }
    }

    private fun landmarker(): PoseLandmarker {
        poseLandmarker?.let { return it }

        val baseOptions = BaseOptions.builder()
            .setModelAssetPath(modelAssetPath)
            .build()
        val options = PoseLandmarker.PoseLandmarkerOptions.builder()
            .setBaseOptions(baseOptions)
            .setRunningMode(RunningMode.VIDEO)
            .setNumPoses(1)
            .setMinPoseDetectionConfidence(minimumPoseDetectionConfidence)
            .setMinPosePresenceConfidence(minimumPosePresenceConfidence)
            .setMinTrackingConfidence(minimumTrackingConfidence)
            .build()
        return PoseLandmarker.createFromOptions(context, options).also {
            poseLandmarker = it
        }
    }

    private fun nextTimestamp(timestampMs: Long): Long {
        val safeTimestampMs = max(timestampMs, lastTimestampMs + 1)
        lastTimestampMs = safeTimestampMs
        return safeTimestampMs
    }

    private fun nv21ToBitmap(bytes: ByteArray, width: Int, height: Int): Bitmap {
        val image = YuvImage(bytes, ImageFormat.NV21, width, height, null)
        val output = ByteArrayOutputStream()
        image.compressToJpeg(Rect(0, 0, width, height), jpegQuality, output)
        val jpegBytes = output.toByteArray()
        return BitmapFactory.decodeByteArray(jpegBytes, 0, jpegBytes.size)
            ?: throw IllegalArgumentException("Could not decode the camera frame.")
    }

    private fun rotateBitmap(source: Bitmap, rotationDegrees: Int): Bitmap {
        val normalizedRotation = ((rotationDegrees % 360) + 360) % 360
        if (normalizedRotation == 0) {
            return source
        }

        val matrix = Matrix().apply {
            postRotate(normalizedRotation.toFloat())
        }
        return Bitmap.createBitmap(source, 0, 0, source.width, source.height, matrix, true)
    }

    private fun resizeBitmapIfNeeded(source: Bitmap): Bitmap {
        val longEdge = max(source.width, source.height)
        if (longEdge <= maxAnalysisLongEdgePx) {
            return source
        }

        val scale = maxAnalysisLongEdgePx.toFloat() / longEdge.toFloat()
        val targetWidth = max(1, (source.width * scale).roundToInt())
        val targetHeight = max(1, (source.height * scale).roundToInt())
        return Bitmap.createScaledBitmap(source, targetWidth, targetHeight, true)
    }

    private fun optionalFloat(value: Optional<Float>): Float? =
        if (value.isPresent) value.get() else null

    companion object {
        private const val channelName = "football_note/mediapipe_pose_landmarker"
        private const val detectMethodName = "detectPoseFromNv21"
        private const val closeMethodName = "close"
        private const val modelAssetPath = "pose_landmarker_lite.task"
        private const val minimumPoseDetectionConfidence = 0.45f
        private const val minimumPosePresenceConfidence = 0.45f
        private const val minimumTrackingConfidence = 0.45f
        private const val jpegQuality = 95
        private const val maxAnalysisLongEdgePx = 720
    }
}

File: ios/Runner/MediaPipePoseLandmarkerChannel.swift
import Flutter
import MediaPipeTasksVision
import UIKit

final class MediaPipePoseLandmarkerChannel {
  private let channel: FlutterMethodChannel
  private let queue = DispatchQueue(
    label: "com.namsoon.footballnote.mediapipe-pose-landmarker",
    qos: .userInitiated
  )
  private var poseLandmarker: PoseLandmarker?
  private var lastTimestampMs = 0

  init(binaryMessenger: FlutterBinaryMessenger) {
    channel = FlutterMethodChannel(
      name: Self.channelName,
      binaryMessenger: binaryMessenger
    )
    channel.setMethodCallHandler { [weak self] call, result in
      self?.handle(call: call, result: result)
    }
  }

  deinit {
    channel.setMethodCallHandler(nil)
  }

  private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case Self.detectBgraMethodName:
      detectPoseFromBgra8888(call: call, result: result)
    case Self.closeMethodName:
      close(result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func detectPoseFromBgra8888(
    call: FlutterMethodCall,
    result: @escaping FlutterResult
  ) {
    guard
      let arguments = call.arguments as? [String: Any],
      let typedData = arguments["bytes"] as? FlutterStandardTypedData,
      let width = arguments["width"] as? Int,
      let height = arguments["height"] as? Int,
      let bytesPerRow = arguments["bytesPerRow"] as? Int,
      width > 0,
      height > 0,
      bytesPerRow >= width * Self.bytesPerPixel
    else {
      result(
        FlutterError(
          code: "invalid_frame",
          message: "MediaPipe pose detection requires BGRA bytes with width, height, and bytesPerRow.",
          details: nil
        )
      )
      return
    }

    let rotationDegrees = arguments["rotationDegrees"] as? Int ?? 0
    let timestampMs = (arguments["timestampMs"] as? NSNumber)?.intValue ?? 0
    let frameData = typedData.data

    queue.async {
      do {
        let detection = try self.detectPose(
          data: frameData,
          width: width,
          height: height,
          bytesPerRow: bytesPerRow,
          rotationDegrees: rotationDegrees,
          timestampMs: timestampMs
        )
        DispatchQueue.main.async {
          result(detection)
        }
      } catch {
        DispatchQueue.main.async {
          result(
            FlutterError(
              code: "mediapipe_pose_failed",
              message: error.localizedDescription,
              details: nil
            )
          )
        }
      }
    }
  }

  private func close(result: @escaping FlutterResult) {
    queue.async {
      self.poseLandmarker = nil
      self.lastTimestampMs = 0
      DispatchQueue.main.async {
        result(nil)
      }
    }
  }

  private func detectPose(
    data: Data,
    width: Int,
    height: Int,
    bytesPerRow: Int,
    rotationDegrees: Int,
    timestampMs: Int
  ) throws -> [String: Any] {
    let image = try makeImage(
      data: data,
      width: width,
      height: height,
      bytesPerRow: bytesPerRow,
      rotationDegrees: rotationDegrees
    )
    let mpImage = try MPImage(uiImage: image)
    let result = try landmarker().detect(
      videoFrame: mpImage,
      timestampInMilliseconds: nextTimestamp(timestampMs)
    )

    return encode(result: result, imageSize: image.size)
  }

  private func landmarker() throws -> PoseLandmarker {
    if let poseLandmarker {
      return poseLandmarker
    }

    guard let modelPath = Bundle.main.path(
      forResource: Self.modelResourceName,
      ofType: Self.modelResourceExtension
    ) else {
      throw MediaPipePoseLandmarkerError.modelMissing
    }

    let options = PoseLandmarkerOptions()
    options.baseOptions.modelAssetPath = modelPath
    options.runningMode = .video
    options.numPoses = 1
    options.minPoseDetectionConfidence = Self.minimumPoseDetectionConfidence
    options.minPosePresenceConfidence = Self.minimumPosePresenceConfidence
    options.minTrackingConfidence = Self.minimumTrackingConfidence

    let createdLandmarker = try PoseLandmarker(options: options)
    poseLandmarker = createdLandmarker
    return createdLandmarker
  }

  private func makeImage(
    data: Data,
    width: Int,
    height: Int,
    bytesPerRow: Int,
    rotationDegrees: Int
  ) throws -> UIImage {
    guard data.count >= bytesPerRow * height else {
      throw MediaPipePoseLandmarkerError.invalidFrame
    }

    guard let provider = CGDataProvider(data: data as CFData) else {
      throw MediaPipePoseLandmarkerError.invalidFrame
    }

    let bitmapInfo = CGBitmapInfo.byteOrder32Little.union(
      CGBitmapInfo(rawValue: CGImageAlphaInfo.premultipliedFirst.rawValue)
    )
    guard let cgImage = CGImage(
      width: width,
      height: height,
      bitsPerComponent: 8,
      bitsPerPixel: 32,
      bytesPerRow: bytesPerRow,
      space: CGColorSpaceCreateDeviceRGB(),
      bitmapInfo: bitmapInfo,
      provider: provider,
      decode: nil,
      shouldInterpolate: false,
      intent: .defaultIntent
    ) else {
      throw MediaPipePoseLandmarkerError.invalidFrame
    }

    return resizedImageIfNeeded(
      rotatedImage(from: cgImage, rotationDegrees: rotationDegrees)
    )
  }

  private func rotatedImage(from cgImage: CGImage, rotationDegrees: Int) -> UIImage {
    let normalizedRotation = ((rotationDegrees % 360) + 360) % 360
    let sourceSize = CGSize(width: cgImage.width, height: cgImage.height)
    guard normalizedRotation != 0 else {
      return UIImage(cgImage: cgImage)
    }

    let outputSize = normalizedRotation == 90 || normalizedRotation == 270
      ? CGSize(width: sourceSize.height, height: sourceSize.width)
      : sourceSize
    let renderer = UIGraphicsImageRenderer(size: outputSize)
    return renderer.image { context in
      let cgContext = context.cgContext
      cgContext.translateBy(x: outputSize.width / 2, y: outputSize.height / 2)
      cgContext.rotate(by: CGFloat(normalizedRotation) * .pi / 180)
      UIImage(cgImage: cgImage).draw(
        in: CGRect(
          x: -sourceSize.width / 2,
          y: -sourceSize.height / 2,
          width: sourceSize.width,
          height: sourceSize.height
        )
      )
    }
  }

  private func resizedImageIfNeeded(_ image: UIImage) -> UIImage {
    let longEdge = max(image.size.width, image.size.height)
    guard longEdge > Self.maxAnalysisLongEdge else {
      return image
    }

    let scale = Self.maxAnalysisLongEdge / longEdge
    let targetSize = CGSize(
      width: max(1, image.size.width * scale),
      height: max(1, image.size.height * scale)
    )
    let renderer = UIGraphicsImageRenderer(size: targetSize)
    return renderer.image { _ in
      image.draw(in: CGRect(origin: .zero, size: targetSize))
    }
  }

  private func nextTimestamp(_ timestampMs: Int) -> Int {
    let safeTimestampMs = max(timestampMs, lastTimestampMs + 1)
    lastTimestampMs = safeTimestampMs
    return safeTimestampMs
  }

  private func encode(result: PoseLandmarkerResult, imageSize: CGSize) -> [String: Any] {
    let landmarks = result.landmarks.first ?? []
    let worldLandmarks = result.worldLandmarks.first ?? []
    let encodedLandmarks = landmarks.enumerated().map { index, landmark in
      let worldLandmark = index < worldLandmarks.count ? worldLandmarks[index] : nil
      var encoded: [String: Any] = [
        "index": index,
        "x": Double(landmark.x) * Double(imageSize.width),
        "y": Double(landmark.y) * Double(imageSize.height),
        "z": Double(landmark.z),
      ]
      if let visibility = optionalDouble(landmark.visibility) {
        encoded["visibility"] = visibility
      }
      if let presence = optionalDouble(landmark.presence) {
        encoded["presence"] = presence
      }
      if let worldLandmark {
        encoded["worldX"] = Double(worldLandmark.x)
        encoded["worldY"] = Double(worldLandmark.y)
        encoded["worldZ"] = Double(worldLandmark.z)
        if let worldVisibility = optionalDouble(worldLandmark.visibility) {
          encoded["worldVisibility"] = worldVisibility
        }
      }
      return encoded
    }

    return [
      "imageWidth": imageSize.width,
      "imageHeight": imageSize.height,
      "landmarks": encodedLandmarks,
    ]
  }

  private func optionalDouble(_ value: NSNumber?) -> Double? {
    value?.doubleValue
  }

  private enum MediaPipePoseLandmarkerError: LocalizedError {
    case invalidFrame
    case modelMissing

    var errorDescription: String? {
      switch self {
      case .invalidFrame:
        return "Could not decode the camera frame."
      case .modelMissing:
        return "MediaPipe pose model is missing from the iOS bundle."
      }
    }
  }

  private static let channelName = "football_note/mediapipe_pose_landmarker"
  private static let detectBgraMethodName = "detectPoseFromBgra8888"
  private static let closeMethodName = "close"
  private static let modelResourceName = "pose_landmarker_lite"
  private static let modelResourceExtension = "task"
  private static let minimumPoseDetectionConfidence: Float = 0.35
  private static let minimumPosePresenceConfidence: Float = 0.35
  private static let minimumTrackingConfidence: Float = 0.35
  private static let bytesPerPixel = 4
  private static let maxAnalysisLongEdge: CGFloat = 720
}
